src/Build/BackEnd/TaskExecutionHost/TaskExecutionHost.cs

```diff
@@ -1169,6 +1173,11 @@ private bool InitializeTaskScalarParameter
                     if (finalTaskItems.Count == 0)
                     {
                         success = true;
+                        if (allowEmptyString)
+                        {
+                            success = SetTaskItemParameter(parameter, new TaskItem(parameterValue, parameterLocation.LocationString, allowEmptyString));
```

Comment:
I think this is subtly wrong. If you have something like `realValue;`, then it would treat that as one item and be ok with it, but it should really interpret that as two values, one of which is empty, and fail.

Reply:
If this is realValue, finalTaskItems.Count will not be zero.  It will  go to the else condition https://github.com/dotnet/msbuild/blob/762ae6c6bbeb21471d4d685a061ed05b47cfcb05/src/Build/BackEnd/TaskExecutionHost/TaskExecutionHost.cs#L1193

Reply:
Correct, but if finalTaskItems.Count is 1, that could mean that finalTaskItems really has one thing in it, or it could have two, where one of them was empty. The bug here is in failing to throw an error when you should.

Reply:
I'm thinking if I should set empty parameter value in the following function when `expression.Length == 0 `before and after expand metadata and properties. That should  throw the expected errors.
https://github.com/dotnet/msbuild/blob/1ae10ed9b7c1a24ff4fa119691e71cfa2975d848/src/Build/Evaluation/Expander.cs#L504-L523